File: internal/apiserver/onboarding.go
package apiserver

import (
	"context"
	"encoding/json"
	"errors"
	"net/http"
	"sync"

	"sigs.k8s.io/controller-runtime/pkg/client"

	ksquadv1 "github.com/K8squad/K8squad/api/v1alpha1"
)

// ============================================================================
// E1 onboarding-progress read model (ISI-3673, AD-2 / FR-1.6) — the SERVER-TRUTH
// projection the Launchpad (E1-S2) and the "Finish setup (n/4)" chip (E1-S3)
// render against:
//
//	GET /api/onboarding/progress → OnboardingProgress{step, done, total: 4, nextMilestone, dismissed}
//
// ============================================================================
//
// AD-2: progress is DERIVED from CRD existence in the caller's Team namespace — no parallel
// progress store (a second store would drift from reality, NFR-5). The four milestones:
//
//	① team    — the Team CR exists (the caller's AuthorContext.TeamID resolves)
//	② agents  — the three preset Roles (role-boss / role-implementer / role-manager, AD-3 /
//	            E2-S1 seeds) are each covered by ≥1 Agent in the namespace
//	③ models  — every Agent CARRIES a credentialSecretRef (non-empty name). Secret-object
//	            existence is deliberately NOT checked here: the apiserver SA has no Secret
//	            read RBAC by design (ISI-3546) and admission-time resolution ("an Agent must
//	            resolve its credential Secret before it is admitted", agent_types.go) is the
//	            webhook/reconciler's fail-closed job (validator.go refExists). The AD-7
//	            test-connection flag refines the milestone: a RECORDED failure un-completes
//	            it; no record means "never tested" and does not block (the E3-S2
//	            test-connection writer lands after E1; a hard requirement would strand the
//	            Launchpad at <4/4 until then)
//	④ project — a Project CR in the namespace has spec.repo.auth.credentialSecretRef set
//
// The only state NOT derivable from CR existence — "user dismissed the Launchpad" and the AD-7
// per-agent test-connection last-result — lives as annotations on the Team CR
// (ksquad.io/onboarding-*), read through the helpers below. The derived step/done counts are
// ALWAYS authoritative over any dismissal flag (AC3): dismissal is a UI convenience, never
// progress.
//
// Tenancy: the route carries NO {teamId} path param — the scope is the session's
// AuthorContext.TeamID, so a cross-tenant read is structurally impossible (AC4; existence-hiding
// by construction, never 403). A MISSING Team CR is not an error here: it is milestone ①
// incomplete — the first-run tenant's honest answer is {step:1, done:0, nextMilestone:"team"},
// not the 404 the other read models give (their projections are meaningless without a Team;
// this projection's whole job is to describe that absence).
// Onboarding milestone IDs, in journey order. nextMilestone carries one of these (empty when
// the journey is complete).
const (
	OnboardingMilestoneTeam    = "team"
	OnboardingMilestoneAgents  = "agents"
	OnboardingMilestoneModels  = "models"
	OnboardingMilestoneProject = "project"
)

// OnboardingTotalMilestones is the fixed journey length (FR-1: the "n/4" denominator).
const OnboardingTotalMilestones = 4

// onboardingPresetRoles are the seeded Role names (AD-3 / E2-S1: config/roles/role-{boss,
// implementer,manager}.yaml, console/lib/presets.ts) whose coverage completes milestone ②.
var onboardingPresetRoles = []string{"role-boss", "role-implementer", "role-manager"}

// Team-annotation keys for the non-CR-derivable flags (AD-2). All live on the Team CR under the
// ksquad.io/onboarding-* prefix.
const (
	// OnboardingDismissedAnnotation marks "user dismissed the Launchpad" ("true"/absent).
	// UI convenience only — the derived step count stays authoritative (AC3).
	OnboardingDismissedAnnotation = "ksquad.io/onboarding-dismissed"
	// onboardingTestConnectionPrefix prefixes the AD-7 per-agent last test-connection
	// result: ksquad.io/onboarding-test-connection-<agentName> = "passed" | "failed".
	onboardingTestConnectionPrefix = "ksquad.io/onboarding-test-connection-"
)

// OnboardingProgress is the projection payload (AC2). Step is the 1-based first INCOMPLETE
// milestone (== Total when the journey is complete); Done counts every complete milestone
// (a tenant who created a Project before their Agents has done=2 even while step=2);
// NextMilestone is the ID of that first incomplete milestone, empty at 4/4. Dismissed surfaces
// the Team-annotation dismissal flag so the Launchpad can honor it without a second read.
type OnboardingProgress struct {
	Step          int    `json:"step"`
	Done          int    `json:"done"`
	Total         int    `json:"total"`
	NextMilestone string `json:"nextMilestone,omitempty"`
	Dismissed     bool   `json:"dismissed"`
}

// OnboardingReader projects onboarding progress for a single Team (identified by its K8s object
// UID, the value AuthorContext.TeamID carries). Mirrors OrgReader (org.go): Team-scoped,
// read-only, cache-backed in prod, fake client in tests.
type OnboardingReader interface {
	// Progress derives the 4-milestone projection for teamUID. A teamUID that resolves to no
	// Team CR returns the zero-progress projection (milestone ① incomplete), NOT an error.
	Progress(ctx context.Context, teamUID string) (OnboardingProgress, error)
}

// ClientOnboardingReader is the production OnboardingReader over any client.Reader (the informer
// cache in the host; a fake client in tests). Team UID→(name, namespace) is memoized on the same
// immutability discipline as ClientOrgReader.
type ClientOnboardingReader struct {
	reader client.Reader
	mu     sync.RWMutex
	teams  map[string]teamIdentity // teamUID → identity (immutable once resolved)
}

// NewClientOnboardingReader builds the onboarding read model over a client.Reader whose scheme
// has api/v1alpha1 registered (the informer cache in prod; a fake client in tests).
func NewClientOnboardingReader(r client.Reader) *ClientOnboardingReader {
	return &ClientOnboardingReader{reader: r, teams: map[string]teamIdentity{}}
}

// resolveTeam mirrors ClientOrgReader.resolveTeam (UID → name/namespace, memoized).
func (r *ClientOnboardingReader) resolveTeam(ctx context.Context, teamUID string) (teamIdentity, error) {
	if teamUID == "" {
		return teamIdentity{}, ErrTeamNotFound
	}
	r.mu.RLock()
	id, hit := r.teams[teamUID]
	r.mu.RUnlock()
	if hit {
		return id, nil
	}
	r.mu.Lock()
	defer r.mu.Unlock()
	if id, ok := r.teams[teamUID]; ok {
		return id, nil
	}
	var teams ksquadv1.TeamList
	if err := r.reader.List(ctx, &teams); err != nil {
		return teamIdentity{}, err
	}
	for i := range teams.Items {
		if string(teams.Items[i].UID) == teamUID {
			id := teamIdentity{name: teams.Items[i].Name, ns: teams.Items[i].Namespace}
			r.teams[teamUID] = id
			return id, nil
		}
	}
	return teamIdentity{}, ErrTeamNotFound
}

// Progress derives the projection. A missing Team is the zero-progress answer, not an error —
// the first-run tenant IS the primary audience of this endpoint.
func (r *ClientOnboardingReader) Progress(ctx context.Context, teamUID string) (OnboardingProgress, error) {
	team, err := r.resolveTeam(ctx, teamUID)
	if errors.Is(err, ErrTeamNotFound) {
		return OnboardingProgress{
			Step:          1,
			Done:          0,
			Total:         OnboardingTotalMilestones,
			NextMilestone: OnboardingMilestoneTeam,
		}, nil
	}
	if err != nil {
		return OnboardingProgress{}, err
	}
	var agents ksquadv1.AgentList
	if err := r.reader.List(ctx, &agents, client.InNamespace(team.ns)); err != nil {
		return OnboardingProgress{}, err
	}
	var projects ksquadv1.ProjectList
	if err := r.reader.List(ctx, &projects, client.InNamespace(team.ns)); err != nil {
		return OnboardingProgress{}, err
	}
	// The dismissal/test-connection flags live on the Team CR — fetch it for the annotations
	// (the resolveTeam list is memoized per UID and does not retain the object).
	var teamCR ksquadv1.Team
	if err := r.reader.Get(ctx, client.ObjectKey{Namespace: team.ns, Name: team.name}, &teamCR); err != nil {
		return OnboardingProgress{}, err
	}
	complete := [OnboardingTotalMilestones]bool{
		true, // ① team — resolved above
		agentsMilestoneComplete(agents.Items),
		modelsMilestoneComplete(agents.Items, &teamCR),
		projectMilestoneComplete(projects.Items),
	}
	return foldProgress(complete, OnboardingDismissed(&teamCR)), nil
}

// foldProgress reduces the four completion booleans into the payload: Done counts all complete
// milestones; Step/NextMilestone name the FIRST incomplete one (the journey's next action).
func foldProgress(complete [OnboardingTotalMilestones]bool, dismissed bool) OnboardingProgress {
	ids := [OnboardingTotalMilestones]string{
		OnboardingMilestoneTeam, OnboardingMilestoneAgents, OnboardingMilestoneModels, OnboardingMilestoneProject,
	}
	p := OnboardingProgress{Total: OnboardingTotalMilestones, Dismissed: dismissed}
	for i, ok := range complete {
		if ok {
			p.Done++
		} else if p.Step == 0 {
			p.Step = i + 1
			p.NextMilestone = ids[i]
		}
	}
	if p.Step == 0 {
		p.Step = OnboardingTotalMilestones // 4/4: parked on the final milestone, no next
	}
	return p
}

// agentsMilestoneComplete (②) requires each of the three seeded preset Roles (AD-3) to be
// covered by ≥1 Agent — which also implies the "≥3 Agents" floor.
func agentsMilestoneComplete(agents []ksquadv1.Agent) bool {
	covered := map[string]bool{}
	for i := range agents {
		covered[agents[i].Spec.RoleRef.Name] = true
	}
	for _, preset := range onboardingPresetRoles {
		if !covered[preset] {
			return false
		}
	}
	return true
}

// modelsMilestoneComplete (③): every Agent carries a credentialSecretRef, and no Agent carries
// a RECORDED test-connection failure on the Team CR (AD-7). Secret-object existence is the
// admission webhook's fail-closed job (the apiserver SA has no Secret RBAC by design,
// ISI-3546), so "resolvable" here means "the ref is set on an admitted Agent". Vacuous truth is
// rejected — with zero Agents the milestone is not done (a tenant with no Agents has not "done"
// models & credentials; milestone ② gates the journey anyway).
func modelsMilestoneComplete(agents []ksquadv1.Agent, team *ksquadv1.Team) bool {
	if len(agents) == 0 {
		return false
	}
	for i := range agents {
		a := &agents[i]
		if a.Spec.CredentialSecretRef.Name == "" {
			return false
		}
		if recorded, passed := TestConnectionFlag(team, a.Name); recorded && !passed {
			return false
		}
	}
	return true
}

// projectMilestoneComplete (④): a Project CR in the namespace has spec.repo.auth set with a
// named credential Secret (project_types.go RepoAuth).
func projectMilestoneComplete(projects []ksquadv1.Project) bool {
	for i := range projects {
		auth := projects[i].Spec.Repo.Auth
		if auth != nil && auth.CredentialSecretRef.Name != "" {
			return true
		}
	}
	return false
}

// ── Team-annotation read/write helpers (AC3 / story task 4) ────────────────
// Pure object-level helpers: they mutate/read the in-memory Team's annotations; persisting a
// write is the caller's applier.Update (the compose-crd write surface owns CR writes, R6).
// OnboardingDismissed reads the Launchpad-dismissal flag from the Team CR.
func OnboardingDismissed(team *ksquadv1.Team) bool {
	return team.Annotations[OnboardingDismissedAnnotation] == "true"
}

// SetOnboardingDismissed sets (or clears, when false) the Launchpad-dismissal flag on the Team
// CR. The caller persists with its writer client.
func SetOnboardingDismissed(team *ksquadv1.Team, dismissed bool) {
	setOnboardingAnnotation(team, OnboardingDismissedAnnotation, dismissed, "true")
}

// TestConnectionFlag reads the AD-7 last test-connection result for agentName from the Team CR:
// (false, false) when no test was ever recorded, otherwise the recorded outcome.
func TestConnectionFlag(team *ksquadv1.Team, agentName string) (recorded, passed bool) {
	switch team.Annotations[onboardingTestConnectionPrefix+agentName] {
	case "passed":
		return true, true
	case "failed":
		return true, false
	default:
		return false, false
	}
}

// SetTestConnectionFlag records the AD-7 last test-connection result for agentName on the Team
// CR. The caller persists with its writer client.
func SetTestConnectionFlag(team *ksquadv1.Team, agentName string, passed bool) {
	setOnboardingAnnotation(team, onboardingTestConnectionPrefix+agentName, true, map[bool]string{true: "passed", false: "failed"}[passed])
}

// setOnboardingAnnotation writes a ksquad.io/onboarding-* annotation, allocating the map when
// needed; when present is false the key is removed (an explicit "absent" beats a stale "false").
func setOnboardingAnnotation(team *ksquadv1.Team, key string, present bool, value string) {
	if !present {
		delete(team.Annotations, key)
		return
	}
	if team.Annotations == nil {
		team.Annotations = map[string]string{}
	}
	team.Annotations[key] = value
}

// onboardingProgress is the handler behind GET /api/onboarding/progress. Team-scoped via the
// session (authScope); a missing Team is milestone ① incomplete, not an error. A read-model
// failure answers 502 (same discipline as the org handlers).
func (s *Server) onboardingProgress(reader OnboardingReader) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		teamUID, ok := authScope(w, r)
		if !ok {
			return
		}
		progress, err := reader.Progress(r.Context(), teamUID)
		if err != nil {
			writeJSONError(w, http.StatusBadGateway, "onboarding read model unavailable")
			return
		}
		writeJSON(w, http.StatusOK, progress)
	}
}

// onboardingDismiss is the handler behind POST /api/onboarding/dismiss (E1-S4 / FR-1.3): the
// server-side write-path that persists the Launchpad-dismissal flag so the "Finish setup (n/4)"
// chip (E1-S3) follows a returning tenant across devices. The GET projection (onboardingProgress)
// only READS the flag; this is its sole writer (BFF stays GET-only by design).
//
// Tenancy mirrors the compose write surface (composecrd.go): the caller's Team is resolved from
// the session (authScope → Team UID), NOT a path param, so a cross-tenant write is structurally
// impossible (AC1). A first-run tenant with no Team CR 404s — dismissal is meaningless before
// milestone ① (AC2). The write rides sameOriginGuard + maxBytesBody at the route (server.go),
// matching the E2-S2 AC3 write conventions.
func (s *Server) onboardingDismiss(applier CRDApplier) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		teamUID, ok := authScope(w, r)
		if !ok {
			return
		}
		var req struct {
			Dismissed bool `json:"dismissed"`
		}
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			writeJSONError(w, http.StatusBadRequest, "invalid JSON body")
			return
		}
		// Resolve the caller's Team CR by UID (the same List-and-match the compose surface uses;
		// authScope hands us the K8s object UID, not a name/namespace). A UID that matches no Team
		// is the first-run tenant — 404 (AC2).
		team, err := resolveTeamByUID(r.Context(), applier, teamUID)
		if err != nil {
			if errors.Is(err, ErrTeamNotFound) {
				writeJSONError(w, http.StatusNotFound, "no team found for this caller")
				return
			}
			writeJSONError(w, http.StatusBadGateway, "team lookup failed")
			return
		}
		SetOnboardingDismissed(team, req.Dismissed)
		if err := applier.Update(r.Context(), team); err != nil {
			writeJSONError(w, http.StatusInternalServerError, "failed to persist dismissal state")
			return
		}
		writeJSON(w, http.StatusOK, map[string]bool{"dismissed": req.Dismissed})
	}
}

// resolveTeamByUID returns the caller's Team CR (the object to mutate + Update) by matching the
// session's Team UID against every Team the applier can see. ErrTeamNotFound when none matches —
// the first-run tenant. Mirrors ComposeService.teamNamespace, but returns the object itself since
// the dismissal write mutates the Team CR's annotations in place.
func resolveTeamByUID(ctx context.Context, applier CRDApplier, teamUID string) (*ksquadv1.Team, error) {
	if teamUID == "" {
		return nil, ErrTeamNotFound
	}
	var teams ksquadv1.TeamList
	if err := applier.List(ctx, &teams); err != nil {
		return nil, err
	}
	for i := range teams.Items {
		if string(teams.Items[i].UID) == teamUID {
			return &teams.Items[i], nil
		}
	}
	return nil, ErrTeamNotFound
}

File: console/lib/onboarding.ts
// lib/onboarding.ts — the Launchpad hub model (E1-S2, ISI-3674; frames 01/06/07; FR-1, AD-2).
//
// PURE derivation layer over the server-truth onboarding projection
// (GET /api/onboarding/progress, E1-S1 / internal/apiserver/onboarding.go):
//
//   { step, done, total: 4, nextMilestone?, dismissed? }
//
// The four milestones, in journey order (AD-2):
//   ① team    — the Team CR exists
//   ② agents  — the three preset Roles (role-boss / role-implementer / role-manager) covered
//   ③ models  — every Agent carries a credentialSecretRef (no recorded test failure)
//   ④ project — a Project CR has spec.repo.auth set
//
// Everything the Launchpad renders is derived HERE (states, hero copy, resume target) so the
// component stays a thin renderer and every derivation is unit-tested. Milestone copy mirrors
// the ISI-3641 mock frames 01/06/07 verbatim; hrefs for the "Review" affordance point at the
// list surfaces (E1-S3's ONBOARDING_MILESTONE_HREF in lib/nav.ts is the routing table for the
// chip's *work* surfaces — kept separate so the two stories merge cleanly).

/** Onboarding milestone ids, in journey order — mirror of the Go constants (onboarding.go). */
export type OnboardingMilestoneId = "team" | "agents" | "models" | "project";

/** The AD-2 projection payload (E1-S1). `step` is 1-based first INCOMPLETE milestone (== total
 * at 4/4); `done` counts every complete milestone (out-of-order completion counts); `dismissed`
 * surfaces the Team-CR annotation ksquad.io/onboarding-dismissed. */
export type OnboardingProgress = {
  step: number;
  done: number;
  total: number;
  nextMilestone?: string;
  dismissed?: boolean;
};

/** Card state per the mock: frame 01 (0/4) shows every card "To do"; frame 06 (resume) shows
 * done cards "Done", the first incomplete "Up next", and later incompletes "Locked". */
export type MilestoneState = "done" | "next" | "locked" | "todo";

export type OnboardingMilestone = {
  id: OnboardingMilestoneId;
  /** 1-based journey position (matches the projection's `step`). */
  step: number;
  /** Card title (frames 01/06). */
  title: string;
  /** One-line "why" (FR-1.1). */
  why: string;
  /** Small tag line under the why (frame copy). */
  tag: string;
  /** Spine label (short). */
  spine: string;
  /** "Review" affordance target once done — the milestone's LIST surface. */
  reviewHref: string;
  /** Tag shown on the frame-07 summary card. */
  summaryTag: string;
};

/** The four milestones in journey order, copy aligned to mock frames 01/06/07. */
export const ONBOARDING_MILESTONES: ReadonlyArray<OnboardingMilestone> = [
  {
    id: "team",
    step: 1,
    title: "Create a Team",
    why: "A squad is a Team of agents that work as one unit.",
    tag: "Boss · Implementation · Manager",
    spine: "Team",
    reviewHref: "/teams",
    summaryTag: "1 team",
  },
  {
    id: "agents",
    step: 2,
    title: "Add your agents",
    why: "Start from a starter-squad template, then tune.",
    tag: "3 roles, pre-wired",
    spine: "Agents",
    reviewHref: "/agents",
    summaryTag: "preset squad",
  },
  {
    id: "models",
    step: 3,
    title: "Choose models",
    why: "Pick a primary model + a fallback for each agent.",
    tag: "BYO key or shared",
    spine: "Models",
    reviewHref: "/credentials",
    summaryTag: "primary + fallback",
  },
  {
    id: "project",
    step: 4,
    title: "Connect a Project",
    why: "Point a repo + GitHub token so agents can ship.",
    tag: "repo + secret://token",
    spine: "Project",
    reviewHref: "/projects",
    summaryTag: "repo + secret://token",
  },
];

export const ONBOARDING_TOTAL = ONBOARDING_MILESTONES.length;

/** The seeded preset-Role sequence milestone ② asks for (AD-3 / E2-S1 seeds in config/roles/;
 * the BE derives completion from the same three ids — onboarding.go onboardingPresetRoles).
 * Used to guide the template on-ramp's per-agent suggestion; the Role CRs themselves carry no
 * model (F-CRD-1), so no model defaults live here. */
export const SUGGESTED_AGENT_ROLES: ReadonlyArray<{
  roleRef: string;
  label: string;
  summary: string;
}> = [
  { roleRef: "role-boss", label: "Boss", summary: "Plans, decomposes and delegates." },
  { roleRef: "role-implementer", label: "Implementer", summary: "Writes, tests and ships code." },
  { roleRef: "role-manager", label: "Manager", summary: "Grooms, reviews and unblocks." },
];

/** Type guard: a payload is a well-formed projection (fail-open boundary for the BFF fetch). */
export function isOnboardingProgress(v: unknown): v is OnboardingProgress {
  if (typeof v !== "object" || v === null) return false;
  const p = v as Record<string, unknown>;
  return (
    typeof p.step === "number" &&
    typeof p.done === "number" &&
    typeof p.total === "number" &&
    p.total > 0 &&
    p.done >= 0 &&
    p.done <= p.total
  );
}

/** The first incomplete milestone id, or null at 4/4. Validates against the known ids — an
 * unrecognized `nextMilestone` from a newer server degrades to null (complete-looking) rather
 * than routing Nadia to a surface this console doesn't know. */
export function nextMilestoneId(p: OnboardingProgress): OnboardingMilestoneId | null {
  if (p.done >= p.total) return null;
  const id = p.nextMilestone ?? "";
  return (ONBOARDING_MILESTONES as ReadonlyArray<{ id: string }>).some((m) => m.id === id)
    ? (id as OnboardingMilestoneId)
    : // Server said "incomplete" but named no known milestone — fall back to the step number.
      (ONBOARDING_MILESTONES[p.step - 1]?.id ?? null);
}

/** True when the journey is complete (4/4 — the D5 yield condition). */
export function isJourneyComplete(p: OnboardingProgress): boolean {
  return p.total > 0 && p.done >= p.total;
}

/**
 * Per-card state derivation (frames 01/06):
 *  - milestones before `step` (the first incomplete) → "done" (Review affordance)
 *  - virgin tenant (done == 0): every incomplete card is "todo" (frame 01 — nothing is
 *    "locked" yet; the hero on-ramps are the action, not per-card CTAs)
 *  - resume (done > 0): the FIRST incomplete card is "next" (Resume →), later incompletes are
 *    "locked" (frame 06 — the guided journey is sequential)
 *
 * Payload limit (documented, NFR-5-consistent): the projection carries only the done COUNT and
 * the first-incomplete step, not per-milestone booleans — so a milestone completed OUT OF ORDER
 * (e.g. Project created via Compose before Agents) still counts in the "n of 4" meter but its
 * card renders by journey position ("locked"), not as done. The guided flow is sequential by
 * design (frame 06 locks later cards), so only Compose-direct users can observe this.
 */
export function milestoneStates(
  p: OnboardingProgress,
): Record<OnboardingMilestoneId, MilestoneState> {
  const next = nextMilestoneId(p);
  const virgin = p.done === 0;
  const out = {} as Record<OnboardingMilestoneId, MilestoneState>;
  for (const m of ONBOARDING_MILESTONES) {
    if (p.step > m.step || (next === null && p.done >= p.total)) {
      // `step` is the first INCOMPLETE milestone: anything before it is complete. At 4/4
      // (next === null, step === total) every milestone is done.
      out[m.id] = "done";
    } else if (virgin) {
      out[m.id] = "todo";
    } else if (m.id === next) {
      out[m.id] = "next";
    } else {
      out[m.id] = "locked";
    }
  }
  return out;
}

/**
 * Hero copy per frame. done == 0 → frame 01 ("Welcome to K8squad"); 0 < done < total →
 * frame 06 ("Finish setting up your squad"); done == total → frame 07 ("Your squad is ready").
 */
export function heroCopy(p: OnboardingProgress): { title: string; sub: string } {
  if (isJourneyComplete(p)) {
    return {
      title: "Your squad is ready",
      sub: "Your squad is live — agents composed, models wired, and a project connected.",
    };
  }
  if (p.done === 0) {
    return {
      title: "Welcome to K8squad",
      sub: "Your tenant is empty. Let's assemble your first squad — four quick steps.",
    };
  }
  const left = p.total - p.done;
  const headline = p.done * 2 === p.total ? "You're halfway there" : `${p.done} of ${p.total} complete`;
  return {
    title: "Finish setting up your squad",
    sub: `${headline} — ${left} step${left === 1 ? "" : "s"} left. Pick up where you left off.`,
  };
}

/** The resume CTA label for the frame-06 hero ("Resume setup — Choose models →"). */
export function resumeLabel(p: OnboardingProgress): string {
  const next = nextMilestoneId(p);
  const title = ONBOARDING_MILESTONES.find((m) => m.id === next)?.title;
  return title ? `Resume setup — ${title}` : "Resume setup";
}

/**
 * D5 gate: should the Overview route render the Launchpad instead of SquadOverview?
 * Show while the journey is incomplete AND not dismissed (server annotation OR this device's
 * localStorage flag — FR-1.3). Complete journeys always yield to the normal Overview.
 */
export function shouldShowLaunchpad(
  p: OnboardingProgress,
  localDismissed: boolean,
): boolean {
  if (isJourneyComplete(p)) return false;
  return !(p.dismissed === true || localDismissed);
}

// ── Local dismissal (FR-1.3 v1 floor) ────────────────────────────────────────
// The server projection surfaces `dismissed` from the Team-CR annotation, but E1-S1 ships no
// HTTP write-path for it (the SetOnboardingDismissed helper exists; the route does not). Until
// that lands (follow-up), "Skip for now" persists on THIS device via localStorage; the server
// flag is honored first whenever present, so the cross-device flag wins the moment a writer
// exists. Dismissal is a UI convenience, never progress (E1-S1 AC3) — the derived counts stay
// authoritative.

export const ONBOARDING_DISMISS_KEY = "ksquad.onboarding.dismissed";

/** Read this device's dismissal flag (false outside the browser / on storage errors). */
export function readLocalDismissed(storage?: Pick<Storage, "getItem">): boolean {
  try {
    const s = storage ?? (typeof window !== "undefined" ? window.localStorage : undefined);
    return s?.getItem(ONBOARDING_DISMISS_KEY) === "true";
  } catch {
    return false;
  }
}

/** Persist (or clear, when false) this device's dismissal flag. */
export function writeLocalDismissed(
  dismissed: boolean,
  storage?: Pick<Storage, "setItem" | "removeItem">,
): void {
  try {
    const s = storage ?? (typeof window !== "undefined" ? window.localStorage : undefined);
    if (!s) return;
    if (dismissed) s.setItem(ONBOARDING_DISMISS_KEY, "true");
    else s.removeItem(ONBOARDING_DISMISS_KEY);
  } catch {
    /* storage unavailable (private mode quota) — dismissal simply doesn't persist */
  }
}
